Fix out-of-memory error in mmat

mducle opened the pull request:
In order to speed up the calculation of the matrix-matrix multiplication of a stack of matrices with just pure Matlab code, the computation the `mmat` function expands the matrices so that it can just use a single call to `bsxfun` to compute the element-wise multiplication in parallel followed by a single call to `sum`. 

However, this requires very large temporary arrays such that the system very rapidly runs out of memory - for example, multiplying an N×M by a stack M×N×L of L matrices requires two temporary arrays of size N×M×N×L matrices.

This PR changes the code so that if the memory requirements are too much, it instead computes the stacked matrix multiplication using loops which would be slow but at least the calculation will actually run.

RichardWaiteSTFC commented on external/mmat.m:
```diff
@@ -51,9 +52,46 @@
 nA = [size(A),ones(1,nD-nDA)]; nA = nA(dim); 
 nB = [size(B),ones(1,nD-nDB)]; nB = nB(dim);
 
+neededMem = (numel(A)*nB(2) + numel(B)*nA(1)) * 16;
```
Is this always true? I think the memory of required for an array is 
`numel(array)*bytes_per_element`
It seems that by default MATLAB uses double precision types uint64 & int64 with 8 bytes per element, but it is possible to have uint32 & int32 formats have 4 bytes per element - should we check?
Possibly a silly question, but where is the factor of `16` coming from? It looks like `sw_freemem` returns the memory of the maximum possible array allowed in bytes should it be 8 or 4?

RichardWaiteSTFC commented on external/mmat.m:
```diff
@@ -51,9 +52,46 @@
+        end
+    else
+        output_shape = [szA(1) szB(2:end)];
+        if size(A, 3) ~= size(B, 3), error('Extra dimensions do not agree'); end
```
I think this check could go on line 80 (i.e. `elseif size(A, 3) == size(B, 3)` - then throw an error on `else`) - does that make sense?

RichardWaiteSTFC commented on external/mmat.m:
```diff
@@ -51,9 +52,46 @@
 nD = max(nDA,nDB);
 
+if nD == 2 && isdefaultdim
+    C = A * B;
```
Good to have this shortcut I think, perhaps reduces the overhead of calling `mmat` for 2D arrays - even though we tend to avoid it!

RichardWaiteSTFC reviewed (changes requested):
Just one small suggestion and a question about the memory calculation.
The new branch doesn't look like it will be tested even in our performance tests (because previously we would have got out of memory errors) - is it worth adding a quick unit test looping over the size of arrays doing e.g. `A * A^-1 = I`  (possibly repeated over a third dimension for one of the arrays) ?

RichardWaiteSTFC commented:
> ... is it worth adding a quick unit test looping over the size of arrays doing e.g. `A * A^-1 = I` (possibly repeated over a third dimension for one of the arrays) ?

Second thoughts it's hard to guarantee the computer won't have enough ram unless you make this very big and then it might take quite a while...

mducle commented:
> > ... is it worth adding a quick unit test looping over the size of arrays doing e.g. `A * A^-1 = I` (possibly repeated over a third dimension for one of the arrays) ?
> 
> Second thoughts it's hard to guarantee the computer won't have enough ram unless you make this very big and then it might take quite a while...

No it's fine... I just need to mock `sw_freemem` to return a low value so the check automatically passes... I just spotted a bug in the `mock_function` utility so have to fix that first...

wardsimon commented:
It all looks good to me, however can you check with the memory profiler that your estimate for memory usage is correct. 
The implementation of `repmat` & `permute` does use the memory estimated, however I the next line using `sumsym` & `bsxfunsym` does not do calculations in place (matrix C and B are in memory), leading to a doubling of memory usage.

For the test case:
Estimated memory usage = 591kB
`repmat` & `permute` = 612.89 kB used (0 kB freed)
`sumsym` & `bsxfunsym` = 611.92 kB used (64.44 kb freed)

So maybe multiply the memory estimate by 2.

wardsimon commented on external/mmat.m:
```diff
@@ -51,9 +52,48 @@
 nB = [size(B),ones(1,nD-nDB)]; nB = nB(dim);
 
+% Array is double float which is 8 bytes per element
+neededMem = (numel(A)*nB(2) + numel(B)*nA(1)) * 8;
```
See comment as this might need to be multiplied by 2

RichardWaiteSTFC replied:
Ah that's what it was before but we didn't know why! Worth a comment in the code as well I think for future us!

mducle commented:
@wardsimon Thanks for this. I had 16 bytes per element before - I think I got it from another bit of the code but couldn't remember why. I've updated the code to use this now, and also added your explanation.
